pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/flink.apache.org/v1beta1/FlinkDeployment/customizations.yaml

```diff
@@ -0,0 +1,110 @@
+apiVersion: config.karmada.io/v1alpha1
+kind: ResourceInterpreterCustomization
+metadata:
+  name: declarative-configuration-flinkdeployment
+spec:
+  target:
+    apiVersion: flink.apache.org/v1beta1
+    kind: FlinkDeployment
+  customizations:
+    healthInterpretation:
+      luaScript: >
+        function InterpretHealth(observedObj)
+          if observedObj.status ~= nil and observedObj.status.jobStatus ~= nil then
+            return observedObj.status.jobStatus.state ~= 'CREATED' and observedObj.status.jobStatus.state ~= 'RECONCILING'
+          end
+          return false
+        end
+    replicaResource:
+      luaScript: >
+        local kube = require("kube")
+
+        local function isempty(s)
+          return s == nil or s == ''
+        end
+
+        function GetReplicas(observedObj)
+          -- FlinkDeployments presently will not be subdivided among clusters, replica should be 1
+          replica = 1
+          requires = {
+            resourceRequest = {},
+          }
+          -- Add jobmanager resources into replica requirement
+
+          jm_replicas = observedObj.spec.jobManager.replicas
+          if isempty(jm_replicas) then
+            jm_replicas = 1
+          end
+
+          for i = 1, jm_replicas do
+            requires.resourceRequest.cpu = kube.resourceAdd(requires.resourceRequest.cpu, tostring(observedObj.spec.jobManager.resource.cpu))
+            requires.resourceRequest.memory = kube.resourceAdd(requires.resourceRequest.memory, observedObj.spec.jobManager.resource.memory)
+          end
+
+          -- Add task manager resources into replica requirement
+
+          parallelism = observedObj.spec.job.parallelism
+          tms = math.ceil(parallelism / observedObj.spec.flinkConfiguration['taskmanager.numberOfTaskSlots'])
+
+          for i = 1, tms do
+            requires.resourceRequest.cpu = kube.resourceAdd(requires.resourceRequest.cpu, tostring(observedObj.spec.taskManager.resource.cpu))
+            requires.resourceRequest.memory = kube.resourceAdd(requires.resourceRequest.memory, observedObj.spec.taskManager.resource.memory)
+          end
+
+          return replica, requires
+        end
```

Comment:
How are the tolerations and nodeAffinity constraints interpreted from the flinkdeployment podTemplate?

Reply:
FlinkDeployments allow you to define podTemplates in a layered manner - meaning you can have a podTemplate with shared attributes between JobManagers and Taskmanagers, as well as individual podTemplates for both JM and TMs. 

The tolerations and nodeAffinity are something that will be handled by the Flink Operator. 

More info here: https://nightlies.apache.org/flink/flink-kubernetes-operator-docs-release-1.8/docs/custom-resource/pod-template/

Reply:
While scheduling, the resource estimator takes the tolerations and nodeAffinity constraints of the workload into account to determine the number of replicas that can be accommodated within a member cluster (based on the taints etc). I was wondering how that information is interpreted from the FlinkDeployment.

Reply:
I do see that there's a way to provide that eg: https://github.com/karmada-io/karmada/blob/84b971a501ba82c53a5ad455c2fe84d842cd7d4e/pkg/apis/config/v1alpha1/resourceinterpretercustomization_types.go#L138

Reply:
I guess what @bharathguvvala means current PR should parse the toleration and nodeAffinity constraints from the pod template and set to [NodeClaim](https://github.com/karmada-io/karmada/blob/0372cabba6f2e26346319aad8b8600de5acff832/pkg/apis/work/v1alpha2/binding_types.go#L183).

Reply:
Yes @RainbowMango . That's what I meant. Something like this would work.

```
requires.nodeClaim = {}
requires.nodeClaim.nodeSelector = observedObj.spec.podTemplate.spec.nodeSelector
requires.nodeClaim.tolerations = observedObj.spec.podTemplate.spec.tolerations
```

Reply:
Got it, thanks for this information!

Reply:
I think again since we don't support multiple pod templates yet, we can just take the nodeAffinity and tolerations from the common podTemplate. Otherwise we can't set differing nodeAffinities / tolerations for JM and TMs.